### ISSUE TITLE
Disable SSL Medium Strength Cipher Suites Support.

### ISSUE THREAD

**Issue by pashok2398** [kubernetes/kubernetes#41038]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):Yes

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):Cipher,3DES,SHA1

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Disable SSL Medium Strength Cipher Suites Support.

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**:
Bare-Metal

- **OS** (e.g. from /etc/os-release):
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

- **Kernel** (e.g. `uname -a`):
Linux xxx 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

- **Install tools**:
- **Others**:

**What happened**:
Based on our security scanning: The remote service supports the use of medium strength SSL ciphers.
TLSv1
ECDHE-RSA-DES-CBC3-SHA Kx=ECDH Au=RSA Enc=3DES-CBC(168) Mac=SHA1 
DES-CBC3-SHA Kx=RSA Au=RSA Enc=3DES-CBC(168) Mac=SHA1

**What you expected to happen**:
Reconfigure the kube-apiserver to avoid use of medium strength ciphers.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

**Comment by calebamiles** [comment]
Could someone from @kubernetes/sig-auth-feature-requests please take a look? Thanks!

**Comment by liggitt** [comment]
Minimum TLS version is TLS1.2, not sure why TLSv1 was indicated.

The allowed ciphers are the golang defaults

**Comment by liggitt** [comment]
Options:

1. We could modify the allowed ciphers from the golang defaults to a hand-crafted set, though it will likely to reach complete agreement on which ciphers are acceptable (lots of clients with varying TLS capabilities use the API)

2. We could externalize control over allowed ciphers, though http2 enablement makes that ordering really tricky to get right, and the config would be unpleasant to express in flags

**Comment by raesene** [comment]
On this, I'm not sure you'd need to get agreement on an exact set of ciphers to address the issue.  Generally scanners are going to flag up any use of 3DES as an issue, so just dropping support for that would help from a compliance standpoint and realistically there are very few possible clients which can't do better than 3DES 

Testing a sample 1.6 kubeadm cluster with https://github.com/drwetter/testssl.sh shows

```
 Testing 359 ciphers via OpenSSL plus sockets against the server, ordered by encryption strength

Hexcode  Cipher Suite Name (OpenSSL)       KeyExch.   Encryption  Bits     Cipher Suite Name (RFC)
-----------------------------------------------------------------------------------------------------------------------------
 xc030   ECDHE-RSA-AES256-GCM-SHA384       ECDH 256   AESGCM      256      TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
 xc014   ECDHE-RSA-AES256-SHA              ECDH 256   AES         256      TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
 x9d     AES256-GCM-SHA384                 RSA        AESGCM      256      TLS_RSA_WITH_AES_256_GCM_SHA384
 x35     AES256-SHA                        RSA        AES         256      TLS_RSA_WITH_AES_256_CBC_SHA
 xc02f   ECDHE-RSA-AES128-GCM-SHA256       ECDH 256   AESGCM      128      TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
 xc013   ECDHE-RSA-AES128-SHA              ECDH 256   AES         128      TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
 x9c     AES128-GCM-SHA256                 RSA        AESGCM      128      TLS_RSA_WITH_AES_128_GCM_SHA256
 x2f     AES128-SHA                        RSA        AES         128      TLS_RSA_WITH_AES_128_CBC_SHA
 xc012   ECDHE-RSA-DES-CBC3-SHA            ECDH 256   3DES        168      TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
 x0a     DES-CBC3-SHA                      RSA        3DES        168      TLS_RSA_WITH_3DES_EDE_CBC_SHA
```

so realistically in that lot, just dropping the 2 3DES options would address the issue.  Running that tool will show some clients that wouldn't work without 3DES but they're all really old and, I think, unsupported, so very unlikely to cause issues.

**Comment by huornlmj** [comment]
Can anyone tell me how to manually disable the weak ciphers that are being exposed?

**Comment by victorgp** [comment]
Kubernetes is a critical piece in the infrastructure of anyone using it and letting it be vulnerable due to weak ciphers i think is a big issue.
I understand it is not easy to agree what ciphers to set and specially, what order, so if we want to make kubernetes secure, then let's let the users select their own ciphers with their own order.

Yes, it would be an extra parameter in the apiserver and a long one, but i don't think we should compromise the Kubernetes security just because of an extra parameter.

The same applies for the Kubelet.

If you agree this is the approach we should follow i will create a pull requests with the needed changes.

**Comment by huornlmj** [comment]
Yes I agree that users should be able to determine and configure the ciphers to use.

**Comment by victorgp** [comment]
I've made this pull request to address this issue, for the moment only for apiserver: https://github.com/kubernetes/kubernetes/pull/48859 

if you want to test it out, this is a ciphers combination i tested without the weak ones:

```
TLS_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA```

**Comment by victorgp** [comment]
The pull request  i made https://github.com/kubernetes/kubernetes/pull/48859 also includes the Kubelet now

**Comment by meeca1218** [comment]
We also received warning 
Port: 6443 Here is the list of medium strength SSL ciphers supported by the remote server :

Medium Strength Ciphers (> 64-bit and < 112-bit key, or 3DES)

ECDHE-RSA-DES-CBC3-SHA Kx=ECDH Au=RSA Enc=3DES-CBC(168) Mac=SHA1
DES-CBC3-SHA Kx=RSA Au=RSA Enc=3DES-CBC(168) Mac=SHA1 
..............................................

How could we fix it ?  Any suggestion ?
